framework/base/Application.php

```diff
@@ -300,19 +300,24 @@ protected function bootstrap()
             }
         }
 
-        foreach ($this->bootstrap as $class) {
+        foreach ($this->bootstrap as $mixed) {
             $component = null;
-            if (is_string($class)) {
-                if ($this->has($class)) {
-                    $component = $this->get($class);
-                } elseif ($this->hasModule($class)) {
-                    $component = $this->getModule($class);
-                } elseif (strpos($class, '\\') === false) {
-                    throw new InvalidConfigException("Unknown bootstrapping component ID: $class");
+            if ($mixed instanceof \Closure) {
+                if (!$component = call_user_func($mixed, $this)) {
+                    continue;
+                }
```

Comment:
in the closure case there should also be a log like in other cases below.
e.g. `Yii::trace('Bootstrap with Closure', __METHOD__);`